src/coreclr/jit/codegenwasm.cpp

```diff
@@ -59,6 +72,256 @@ void CodeGen::genFuncletEpilog()
+    {
+        case BBJ_RETURN:
+            genExitCode(block);
+            break;
+
+        case BBJ_THROW:
+            // TODO-WASM-CQ: only emit this when needed.
+            instGen(INS_unreachable);
+            break;
+
+        case BBJ_CALLFINALLY:
+            // TODO-WASM: EH model
+            genCallFinally(block);
+            if (!block->isBBCallFinallyPair())
+            {
+                instGen(INS_unreachable);
+            }
+            break;
+
+        case BBJ_EHCATCHRET:
+            // TODO-WASM: EH model
+            genEHCatchRet(block);
+            FALLTHROUGH;
+
+        case BBJ_EHFINALLYRET:
+        case BBJ_EHFAULTRET:
+        case BBJ_EHFILTERRET:
+            // TODO-WASM: EH model
+            genReserveFuncletEpilog(block);
+            break;
+
+        case BBJ_SWITCH:
+        {
+            BBswtDesc* const desc      = block->GetSwitchTargets();
+            unsigned const   caseCount = desc->GetCaseCount();
+
+            assert(!desc->HasDefaultCase());
+
+            if (caseCount == 0)
+            {
+                break;
+            }
+
+            GetEmitter()->emitIns_I(INS_br_table, EA_4BYTE, caseCount);
+
+            for (unsigned caseNum = 0; caseNum < caseCount; caseNum++)
+            {
+                BasicBlock* const caseTarget = desc->GetCase(caseNum)->getDestinationBlock();
+                unsigned          depth      = findTargetDepth(caseTarget);
+
+                GetEmitter()->emitIns_I(INS_label, EA_4BYTE, depth);
+            }
+
+            JITDUMP("\n");
+            break;
+        }
+
+        case BBJ_CALLFINALLYRET:
+        case BBJ_ALWAYS:
+        {
+            BasicBlock* const target = block->GetTarget();
+
+            // If this block jumps to the next one, we might be able to skip emitting the jump
+            //
+            if (block->CanRemoveJumpToNext(compiler))
+            {
+                assert(getBlockIndex(target) == (getBlockIndex(block) + 1));
+                break;
+            }
+
+            inst_JMP(EJ_jmp, target);
+            break;
+        }
+
+        case BBJ_COND:
+        {
+            BasicBlock* const trueTarget  = block->GetTrueTarget();
+            BasicBlock* const falseTarget = block->GetFalseTarget();
+
+            // We don't expect degenerate BBJ_COND
+            //
+            assert(trueTarget != falseTarget);
+
+            // We don't expect the true target to be the next block.
+            //
+            assert(trueTarget != block->Next());
+
+            // br_if for true target
+            //
+            inst_JMP(EJ_jmpif, trueTarget);
+
+            // br for false target, if not fallthrough
+            //
+            if (falseTarget == block->Next())
+            {
+                break;
+            }
+            inst_JMP(EJ_jmp, falseTarget);
+
+            break;
+        }
+
+        default:
+            noway_assert(!"Unexpected bbKind");
+            break;
+    }
```

Comment:
Is it not possible to keep the wasm version of `genEmitEndBlock` unified with the other backends? The differences as far as I can tell arise just from diverging with the other backends in the handling of `GT_SWITCH` and `GT_JTRUE`.

Reply:
There may be more once we solidify the EH design, but maybe those can be folded into the current high-level model as well.